### ISSUE TITLE
Rule to enforce type declaration of `$state` rune

### ISSUE THREAD

**Issue by AntonioVentilii** [sveltejs/eslint-plugin-svelte#1230]
### Motivation

Is there a rule that enforces the declaration of the types for `$state`? If not, is it worth doing?

### Description

For consistency, we could have a rule similar to [prefer-reduce-type-parameter](https://typescript-eslint.io/rules/prefer-reduce-type-parameter/), that clearly enforces where to define the type of a `$state` variable.

### Examples

```svelte
<script>

  // BAD
  let title: string | undefined = $state();
  let data: { value?: number } = $state({})
  
  // GOOD
  let title = $state<string | undefined>();
  let data = $state<{ value?: number }>({});
  
  // UNCHANGED
  let label = $state(0)

</script>
```

**Comment by dimfeld** [comment]
Is there a reason that passing the type as a generic parameter is preferred? The docs suggest the "bad" way: https://svelte.dev/docs/svelte/typescript#Typing-$state

**Comment by baseballyama** [comment]
Maybe we can refer to https://github.com/sveltejs/svelte/pull/10694  to understand why we don’t use generics for runes.

**Comment by marekdedic** [comment]
IMO, Rich's reasoning with inference doesn't apply to `$state` because unlike `$props`, the initializer value is a parameter and therefore TS will error out if it's incorrect.

However, IMHO, it's better to be consistent between `$state` and `$props` - so my preference would actually be a rule that does the exact opposite from the original proposal :D

**Comment by AntonioVentilii** [comment]
Maybe a rule where one can choose either ways? My point was generally just to have consistency in the code, not leaning on a specific one of them, just preferred practices, either. There are already a few rules that are `prefer-*` but they don't mean that they are compulsory, just consistent.

**Comment by WarningImHack3r** [comment]
> Is there a reason that passing the type as a generic parameter is preferred? The docs suggest the "bad" way: [svelte.dev/docs/svelte/typescript#Typing-$state](https://svelte.dev/docs/svelte/typescript#Typing-$state)

I much prefer typing with generics, as you don't get any of the type mismatch you get when specifying the types the way the docs say:
```svelte
<script lang="ts">
	let wrong: number = $state(); // conflict between variable type and `$state` type
    //  ^? number       ^? number | undefined
	let good = $state<number>(); // no conflict: as there is no default value, the type is properly coalesced to "| undefined"
    //  ^? number | undefined
</script>
```

---

Similarly, I've seen some people do this kind of things for `$derived.by`:
```svelte
<script lang="ts">
	const d = $derived.by((): number => {
        return 42; // ignore the useless `.by` instead of a regular `$derived()`
    });
</script>
```
Instead of using generics:
```svelte
<script lang="ts">
	const d = $derived.by<number>(() => {
        return 42;
    });
</script>
```
This rule should also catch with a "cross-runes" name: `prefer-generics-typing`.

> [!NOTE]
> Unlike the `$state` case where generics can make life easier and conflict-free, I could not think of a case where non-generic typing inside `$derived.by` as showcased can be a hassle; it's more a matter of cross-runes consistency and overall best practice when using functions (runes) that support generics.
